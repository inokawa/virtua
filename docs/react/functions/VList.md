[**API**](../../API.md) • **Docs**

***

# Function: VList()

> **VList**(`props`): `ReactNode`

Virtualized list component. See [VListProps](../interfaces/VListProps.md) and [VListHandle](../interfaces/VListHandle.md).

## Parameters

• **props**: [`VListProps`](../interfaces/VListProps.md) & `RefAttributes`\<[`VListHandle`](../interfaces/VListHandle.md)\>

## Returns

`ReactNode`

## Defined in

[src/react/VList.tsx:44](https://github.com/inokawa/virtua/blob/7b801f16c7f1cf5eb033801b816966faaa8a6b18/src/react/VList.tsx#L44)
