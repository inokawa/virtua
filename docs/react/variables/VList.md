[**API**](../../API.md)

***

# Variable: VList

> `const` **VList**: `ForwardRefExoticComponent`\<[`VListProps`](../interfaces/VListProps.md) & `RefAttributes`\<[`VListHandle`](../interfaces/VListHandle.md)\>\>

Defined in: [src/react/VList.tsx:43](https://github.com/inokawa/virtua/blob/71c97bdad291763ca5072b4af388608178a6e6ea/src/react/VList.tsx#L43)

Virtualized list component. See [VListProps](../interfaces/VListProps.md) and [VListHandle](../interfaces/VListHandle.md).
