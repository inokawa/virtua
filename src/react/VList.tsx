import { ReactElement, forwardRef, useRef } from "react";
import { ViewportComponentAttributes } from "./types";
import {
  Virtualizer,
  VirtualizerHandle,
  VirtualizerProps,
} from "./Virtualizer";
import { NULL } from "../core/utils";

/**
 * Methods of {@link VList}.
 */
export interface VListHandle extends VirtualizerHandle {}

/**
 * Props of {@link VList}.
 */
export interface VListProps
  extends Pick<
      VirtualizerProps,
      | "children"
      | "count"
      | "overscan"
      | "itemSize"
      | "shift"
      | "horizontal"
      | "cache"
      | "ssrCount"
      | "item"
      | "onScroll"
      | "onScrollEnd"
      | "keepMounted"
    >,
    ViewportComponentAttributes {
  /**
   * If true, items are aligned to the end of the list when total size of items are smaller than viewport size. It's useful for chat like app.
   */
  reverse?: boolean;
}

/**
 * Virtualized list component. See {@link VListProps} and {@link VListHandle}.
 */
export const VList = forwardRef<VListHandle, VListProps>(
  (
    {
      children,
      count,
      overscan,
      itemSize,
      shift,
      horizontal,
      keepMounted,
      reverse,
      cache,
      ssrCount,
      item,
      onScroll,
      onScrollEnd,
      style,
      ...attrs
    },
    ref
  ): ReactElement => {
    const scrollRef = useRef<HTMLDivElement>(NULL);
    const shouldReverse = reverse && !horizontal;

    let element = (
      <Virtualizer
        ref={ref}
        scrollRef={shouldReverse ? scrollRef : undefined}
        count={count}
        overscan={overscan}
        itemSize={itemSize}
        shift={shift}
        horizontal={horizontal}
        keepMounted={keepMounted}
        cache={cache}
        ssrCount={ssrCount}
        item={item}
        onScroll={onScroll}
        onScrollEnd={onScrollEnd}
      >
        {children}
      </Virtualizer>
    );

    if (shouldReverse) {
      element = (
        <div
          style={{
            visibility: "hidden", // TODO replace with other optimization methods
            display: "flex",
            flexDirection: "column",
            justifyContent: "flex-end",
            minHeight: "100%",
          }}
        >
          {element}
        </div>
      );
    }

    return (
      <div
        ref={scrollRef}
        {...attrs}
        style={{
          display: horizontal ? "inline-block" : "block",
          [horizontal ? "overflowX" : "overflowY"]: "auto",
          contain: "strict",
          width: "100%",
          height: "100%",
          ...style,
        }}
      >
        {element}
      </div>
    );
  }
);
