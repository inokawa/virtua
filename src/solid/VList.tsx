/**
 * @jsxImportSource solid-js
 */
import { JSX } from "solid-js";
import { ViewportComponentAttributes } from "./types";
import {
  Virtualizer,
  VirtualizerHandle,
  VirtualizerProps,
} from "./Virtualizer";

/**
 * Methods of {@link VList}.
 */
export interface VListHandle extends VirtualizerHandle {}

/**
 * Props of {@link VList}.
 */
export interface VListProps<T>
  extends Pick<
      VirtualizerProps<T>,
      | "ref"
      | "data"
      | "children"
      | "overscan"
      | "itemSize"
      | "shift"
      | "horizontal"
      | "onScroll"
      | "onScrollEnd"
    >,
    ViewportComponentAttributes {}

/**
 * Virtualized list component. See {@link VListProps} and {@link VListHandle}.
 */
export const VList = <T,>(props: VListProps<T>): JSX.Element => {
  const {
    ref,
    data,
    children,
    overscan,
    itemSize,
    shift,
    horizontal,
    onScroll,
    onScrollEnd,
    style,
    ...attrs
  } = props;

  return (
    <div
      {...attrs}
      style={{
        display: horizontal ? "inline-block" : "block",
        [horizontal ? "overflow-x" : "overflow-y"]: "auto",
        contain: "strict",
        width: "100%",
        height: "100%",
        ...props.style,
      }}
    >
      <Virtualizer
        ref={props.ref}
        data={props.data}
        overscan={props.overscan}
        itemSize={props.itemSize}
        shift={props.shift}
        horizontal={horizontal}
        onScroll={props.onScroll}
        onScrollEnd={props.onScrollEnd}
      >
        {props.children}
      </Virtualizer>
    </div>
  );
};
