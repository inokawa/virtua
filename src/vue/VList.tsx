/** @jsxImportSource vue */
import {
  defineComponent,
  ComponentOptionsMixin,
  SlotsType,
  ComponentOptionsWithObjectProps,
  ComponentObjectPropsOptions,
  ref,
  VNode,
} from "vue";
import { Virtualizer, VirtualizerHandle } from "./Virtualizer";

interface VListHandle extends VirtualizerHandle {}

const props = {
  /**
   * The data items rendered by this component.
   */
  data: { type: Array, required: true },
  /**
   * Number of items to render above/below the visible bounds of the list. You can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan: { type: Number, default: 4 },
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize: Number,
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift: Boolean,
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal: Boolean,
  /**
   * A prop for SSR. If set, the specified amount of items will be mounted in the initial rendering regardless of the container size until hydrated.
   */
  ssrCount: Number,
} satisfies ComponentObjectPropsOptions;

export const VList = /*#__PURE__*/ defineComponent({
  props: props,
  emits: ["scroll", "scrollEnd", "rangeChange"],
  setup(props, { emit, expose, slots }) {
    const horizontal = props.horizontal;

    const onScroll = (offset: number) => {
      emit("scroll", offset);
    };
    const onScrollEnd = () => {
      emit("scrollEnd");
    };
    const onRangeChange = (start: number, end: number) => {
      emit("rangeChange", start, end);
    };

    const handle = ref<InstanceType<typeof Virtualizer>>();

    expose({
      get scrollOffset() {
        return handle.value!.scrollOffset;
      },
      get scrollSize() {
        return handle.value!.scrollSize;
      },
      get viewportSize() {
        return handle.value!.viewportSize;
      },
      getItemOffset: (...args) => handle.value!.getItemOffset(...args),
      scrollToIndex: (...args) => handle.value!.scrollToIndex(...args),
      scrollTo: (...args) => handle.value!.scrollTo(...args),
      scrollBy: (...args) => handle.value!.scrollBy(...args),
    } satisfies VListHandle);

    return () => {
      return (
        <div
          style={{
            display: horizontal ? "inline-block" : "block",
            [horizontal ? "overflowX" : "overflowY"]: "auto",
            contain: "strict",
            width: "100%",
            height: "100%",
          }}
        >
          <Virtualizer
            ref={handle}
            data={props.data}
            overscan={props.overscan}
            itemSize={props.itemSize}
            shift={props.shift}
            ssrCount={props.ssrCount}
            horizontal={horizontal}
            onScroll={onScroll}
            onScrollEnd={onScrollEnd}
            onRangeChange={onRangeChange}
          >
            {slots}
          </Virtualizer>
        </div>
      );
    };
  },
} as ComponentOptionsWithObjectProps<
  typeof props,
  VListHandle,
  {},
  {},
  {},
  ComponentOptionsMixin,
  ComponentOptionsMixin,
  {
    /**
     * Callback invoked whenever scroll offset changes.
     * @param offset Current scrollTop or scrollLeft.
     */
    scroll: (offset: number) => void;
    /**
     * Callback invoked when scrolling stops.
     */
    scrollEnd: () => void;
    /**
     * Callback invoked when visible items range changes.
     */
    rangeChange: (
      /**
       * The start index of viewable items.
       */
      startIndex: number,
      /**
       * The end index of viewable items.
       */
      endIndex: number
    ) => void;
  },
  string,
  {},
  string,
  SlotsType<{ default: (arg: { item: any; index: number }) => VNode[] }>
>);
