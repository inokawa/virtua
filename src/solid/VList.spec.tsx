/**
 * @jsxImportSource solid-js
 */
import { it, expect, describe, afterEach, vi } from "vitest";
import { render as _render, cleanup } from "@solidjs/testing-library";
import { VList } from "./VList.js";
import { setupResizeJsDom } from "../../scripts/spec.js";
import { type JSX } from "solid-js";

const ITEM_HEIGHT = 50;
const ITEM_WIDTH = 100;
const VIEWPORT_HEIGHT = ITEM_HEIGHT * 10;

setupResizeJsDom({
  itemSize: { width: ITEM_WIDTH, height: ITEM_HEIGHT },
  viewportSize: { width: ITEM_WIDTH, height: VIEWPORT_HEIGHT },
});

const range = (length: number) => Array.from({ length }).map((_, i) => i);

const render = (...args: Parameters<typeof _render>) => {
  const res = _render(...args);
  vi.runAllTicks();
  return res;
};

afterEach(cleanup);

it("should pass attributes to element", () => {
  const { asFragment } = render(() => (
    <VList
      data={range(1)}
      id="id"
      class="class"
      tab-index={0}
      role="list"
      aria-label="test"
      style={{ background: "red" }}
    >
      {(d) => <div>{d}</div>}
    </VList>
  ));
  expect(asFragment()).toMatchSnapshot();
});

it("should render with keepMounted", () => {
  const { asFragment } = render(() => (
    <VList
      data={range(100)}
      keepMounted={[0, 10, 20, 30, 40, 50, 60, 70, 80, 90]}
    >
      {(d) => <div>{d}</div>}
    </VList>
  ));
  expect(asFragment()).toMatchSnapshot();
});

describe("vertical", () => {
  it("should render 0 children", () => {
    const { asFragment } = render(() => (
      <VList data={[]}>{(d) => <div>{d}</div>}</VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 1 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(1)}>{(d) => <div>{d}</div>}</VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 5 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(5)}>{(d) => <div>{d}</div>}</VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 100 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(100)}>{(d) => <div>{d}</div>}</VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 1000 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(1000)}>{(d) => <div>{d}</div>}</VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 10000 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(10000)}>{(d) => <div>{d}</div>}</VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render component", () => {
    const Comp = (props: { children: JSX.Element }) => (
      <div>{props.children}</div>
    );
    const { asFragment } = render(() => (
      <VList data={range(3)}>{(d) => <Comp>{d}</Comp>}</VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with given width / height", () => {
    const { asFragment } = render(() => (
      <VList data={range(5)} style={{ width: "100px", height: "800px" }}>
        {(d) => <div>{d}</div>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("horizontal", () => {
  it("should render 0 children", () => {
    const { asFragment } = render(() => (
      <VList data={[]} horizontal>
        {(d) => <div>{d}</div>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 1 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(1)} horizontal>
        {(d) => <div>{d}</div>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 5 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(5)} horizontal>
        {(d) => <div>{d}</div>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 100 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(100)} horizontal>
        {(d) => <div>{d}</div>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 1000 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(1000)} horizontal>
        {(d) => <div>{d}</div>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 10000 children", () => {
    const { asFragment } = render(() => (
      <VList data={range(10000)} horizontal>
        {(d) => <div>{d}</div>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render component", () => {
    const Comp = (props: { children: JSX.Element }) => (
      <div>{props.children}</div>
    );
    const { asFragment } = render(() => (
      <VList data={range(3)} horizontal>
        {(d) => <Comp>{d}</Comp>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with given width / height", () => {
    const { asFragment } = render(() => (
      <VList
        data={range(5)}
        style={{ width: "100px", height: "800px" }}
        horizontal
      >
        {(d) => <div>{d}</div>}
      </VList>
    ));
    expect(asFragment()).toMatchSnapshot();
    expect(asFragment()).toMatchSnapshot();
  });
});
