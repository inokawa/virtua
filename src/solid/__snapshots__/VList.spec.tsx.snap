// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`horizontal > should render 0 children 1`] = `"<div style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 0px; height: 100%;"></div></div>"`;

exports[`horizontal > should render 1 children 1`] = `"<div style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100px; height: 100%;"><div index="0" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: inline-flex;"><div>0</div></div></div></div>"`;

exports[`horizontal > should render 5 children 1`] = `"<div style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 500px; height: 100%;"><div index="0" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: inline-flex;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: inline-flex;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: inline-flex;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: inline-flex;"><div>3</div></div></div></div>"`;

exports[`horizontal > should render 100 children 1`] = `"<div style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 10000px; height: 100%;"><div index="0" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: inline-flex;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: inline-flex;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: inline-flex;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: inline-flex;"><div>3</div></div></div></div>"`;

exports[`horizontal > should render 10000 children 1`] = `"<div style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 1000000px; height: 100%;"><div index="0" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: inline-flex;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: inline-flex;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: inline-flex;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: inline-flex;"><div>3</div></div></div></div>"`;

exports[`horizontal > should render component 1`] = `"<div style="display: inline-block; overflow-x: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 300px; height: 100%;"><div index="0" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: inline-flex;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: inline-flex;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: inline-flex;"><div>2</div></div></div></div>"`;

exports[`horizontal > should render with given width / height 1`] = `"<div style="display: inline-block; overflow-x: auto; contain: strict; width: 100px; height: 800px;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 500px; height: 100%;"><div index="0" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: inline-flex;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: inline-flex;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: inline-flex;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: inline-flex;"><div>3</div></div></div></div>"`;

exports[`horizontal > should render with given width / height 2`] = `"<div style="display: inline-block; overflow-x: auto; contain: strict; width: 100px; height: 800px;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 500px; height: 100%;"><div index="0" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 0px; visibility: visible; display: inline-flex;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 100px; visibility: visible; display: inline-flex;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 200px; visibility: visible; display: inline-flex;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; height: 100%; top: 0px; left: 300px; visibility: visible; display: inline-flex;"><div>3</div></div></div></div>"`;

exports[`should pass attributes to element 1`] = `"<div id="id" class="class" tab-index="0" role="list" aria-label="test" style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%; background: red;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 50px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div></div></div>"`;

exports[`should render with keepMounted 1`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 5000px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"><div>3</div></div><div index="4" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"><div>4</div></div><div index="5" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 250px; visibility: visible;"><div>5</div></div><div index="6" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 300px; visibility: visible;"><div>6</div></div><div index="7" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 350px; visibility: visible;"><div>7</div></div><div index="8" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 400px; visibility: visible;"><div>8</div></div><div index="9" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 450px; visibility: visible;"><div>9</div></div><div index="10" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 500px; visibility: visible;"><div>10</div></div><div index="11" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 550px; visibility: visible;"><div>11</div></div><div index="12" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 600px; visibility: visible;"><div>12</div></div><div index="13" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 650px; visibility: visible;"><div>13</div></div><div index="14" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 700px; visibility: visible;"><div>14</div></div><div index="20" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 1000px; visibility: visible;"><div>20</div></div><div index="30" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 1500px; visibility: visible;"><div>30</div></div><div index="40" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 2000px; visibility: visible;"><div>40</div></div><div index="50" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 2500px; visibility: visible;"><div>50</div></div><div index="60" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 3000px; visibility: visible;"><div>60</div></div><div index="70" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 3500px; visibility: visible;"><div>70</div></div><div index="80" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 4000px; visibility: visible;"><div>80</div></div><div index="90" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 4500px; visibility: visible;"><div>90</div></div></div></div>"`;

exports[`vertical > should render 0 children 1`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 0px;"></div></div>"`;

exports[`vertical > should render 1 children 1`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 50px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div></div></div>"`;

exports[`vertical > should render 5 children 1`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 250px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"><div>3</div></div><div index="4" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"><div>4</div></div></div></div>"`;

exports[`vertical > should render 100 children 1`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 5000px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"><div>3</div></div><div index="4" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"><div>4</div></div><div index="5" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 250px; visibility: visible;"><div>5</div></div><div index="6" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 300px; visibility: visible;"><div>6</div></div><div index="7" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 350px; visibility: visible;"><div>7</div></div><div index="8" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 400px; visibility: visible;"><div>8</div></div><div index="9" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 450px; visibility: visible;"><div>9</div></div><div index="10" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 500px; visibility: visible;"><div>10</div></div><div index="11" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 550px; visibility: visible;"><div>11</div></div><div index="12" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 600px; visibility: visible;"><div>12</div></div><div index="13" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 650px; visibility: visible;"><div>13</div></div><div index="14" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 700px; visibility: visible;"><div>14</div></div></div></div>"`;

exports[`vertical > should render 10000 children 1`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 500000px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"><div>3</div></div><div index="4" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"><div>4</div></div><div index="5" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 250px; visibility: visible;"><div>5</div></div><div index="6" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 300px; visibility: visible;"><div>6</div></div><div index="7" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 350px; visibility: visible;"><div>7</div></div><div index="8" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 400px; visibility: visible;"><div>8</div></div><div index="9" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 450px; visibility: visible;"><div>9</div></div><div index="10" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 500px; visibility: visible;"><div>10</div></div><div index="11" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 550px; visibility: visible;"><div>11</div></div><div index="12" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 600px; visibility: visible;"><div>12</div></div><div index="13" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 650px; visibility: visible;"><div>13</div></div><div index="14" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 700px; visibility: visible;"><div>14</div></div></div></div>"`;

exports[`vertical > should render component 1`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100%; height: 100%;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 150px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"><div>2</div></div></div></div>"`;

exports[`vertical > should render with given width / height 1`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100px; height: 800px;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 250px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"><div>3</div></div><div index="4" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"><div>4</div></div></div></div>"`;

exports[`vertical > should render with given width / height 2`] = `"<div style="display: block; overflow-y: auto; contain: strict; width: 100px; height: 800px;"><div style="contain: size paint style; overflow-anchor: none; overflow: clip; flex: 0 0 auto; position: relative; width: 100%; height: 250px;"><div index="0" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"><div>0</div></div><div index="1" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"><div>1</div></div><div index="2" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"><div>2</div></div><div index="3" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"><div>3</div></div><div index="4" style="contain: layout style; position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"><div>4</div></div></div></div>"`;
